### ISSUE TITLE
[BUG] MoE buffers dominate memory fragmentation with MFSDP v2 and CUDA graphs

### ISSUE THREAD

**Issue by wujingyue** [NVIDIA/Megatron-LM#7622]
**Describe the bug**

The largest reserved-minus-allocated peak gaps are on the main compute and shared MoE dispatch/combine streams. Allocation history shows that most of the retained memory on both streams is MoE-related. The investigation should focus on MoE buffer lifetimes and handoffs between these streams.

After #7612 consolidates 28 dispatch/combine streams into one, training still shows a **33.70 GiB** reserved-minus-allocated peak gap: **226.15 GiB** reserved after warmup minus **192.45 GiB** allocated over the whole run. Including capture, peak reserved is **228.76 GiB**, giving a **36.31 GiB** gap. Values are maxima across ranks.

Measured actual training with mock data on 64 GB300 GPUs: 128K sequence length, EP=CP=16, SGD, 25 steps with five warmup steps, expandable segments, and training plus optimizer CUDA graphs.

**Reproducer**

Download the three files individually into the same directory:

- [replay.py](https://raw.githubusercontent.com/wujingyue/Megatron-LM/e8ac6620dd7449fa3c8a6f55786c39beac84e18e/replay.py)
- [trace-moe.json.gz](https://raw.githubusercontent.com/wujingyue/Megatron-LM/e8ac6620dd7449fa3c8a6f55786c39beac84e18e/trace-moe.json.gz)
- [trace-compute.json.gz](https://raw.githubusercontent.com/wujingyue/Megatron-LM/e8ac6620dd7449fa3c8a6f55786c39beac84e18e/trace-compute.json.gz)

The traces omit original stream handles and source-file labels; allocation events and expected counters are unchanged.

Two reduced traces replay one allocation stream at a time using only Python and PyTorch, on one GPU:

```bash
PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True python replay.py trace-moe.json.gz
PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True python replay.py trace-compute.json.gz
```

Validated on GB300 with PyTorch `2.13.0a0+8145d630e8.nv26.06`:

| Rank-0 stream | Peak allocated, original and replay | Original peak reserved | Replay peak reserved |
| --- | ---: | ---: | ---: |
| Shared MoE dispatch/combine | 14.598862 GiB | 29.035156 GiB | 29.054688 GiB |
| Main compute | 169.434982 GiB | 183.576172 GiB | 183.576172 GiB |

Compute matches allocated, active, and reserved peaks exactly. MoE matches allocated; reserved is 20 MiB higher and active is 35 MiB lower. Each row combines default and graph pools. Per-stream peaks occur at different times and are not additive.

**Investigation notes**

At the main compute stream's graph-pool allocated peak, **13.953125 GiB** has been freed by the application but is still active in the allocator. Allocation stacks identify:

- **13.125 GiB:** MoE grouped-linear outputs.
- **0.328125 GiB:** MoE preprocessing linear outputs.
- **0.500 GiB:** context-parallel attention buffers.

MoE outputs account for **13.453125 GiB, or 95.5%**, of this stream's **14.084618 GiB** graph-pool peak gap. CP attention contributes only **0.500 GiB (3.6%)**. The remaining **0.131493 GiB** is the difference between peak reserved and peak active. Main compute is therefore another location where MoE buffers are retained, rather than evidence of an unrelated major source of fragmentation. Allocation stacks identify the producers of the memory; the precise cause of each delayed free remains under investigation.

MFSDP v2 itself does not call `record_stream`. `ShortcutMoEBlock` does in `_launch_dispatch`, `_wait_dispatch`, `_launch_combine`, and `_wait_combine` when handing tensors between streams.

The replay uses one allocation stream plus one auxiliary stream that allocates nothing. `record_stream` inside a private `MemPool` approximates the observed deferred frees. It preserves allocation sizes, free requests, pool boundaries, and cache flushes, but does not reconstruct the original dependency graph or eager free-completion timing.

**Expected behavior**

Investigate and reduce retention of MoE buffers on the compute and dispatch/combine streams while preserving stream lifetime correctness and training numerics. Determine which memory can safely be reused during capture; pending frees contribute to the gap, so it is not all proven reusable slack.

**Comment by wujingyue** [comment]
For fragmentation analysis, **`max_active` is the appropriate baseline to compare against `max_reserved`**, because `max_allocated` excludes memory whose free has been requested but which is not yet reusable. PyTorch exposes it as `torch.cuda.memory_stats()["active_bytes.all.peak"]`.

MoE uses `record_stream` in [`ShortcutMoEBlock._launch_dispatch`, `_wait_dispatch`, `_launch_combine`, and `_wait_combine`](https://github.com/NVIDIA/Megatron-LM/blob/565de549ec86478b1057ca795bcda8d8b70b8aa2/megatron/core/models/hybrid/shortcut_block.py#L248-L289). MFSDP v2 itself does not call it. This creates a window between deleting the last tensor reference and the allocator making its storage reusable:

- `free_requested` (`f`): current **allocated** decreases.
- `free_completed` (`c`): current **active** decreases once the allocator can safely reuse the block.

`record_stream` can delay the second event while the first still happens immediately. Thus, `active = allocated + pending-free memory`, and a reserved-minus-allocated gap includes pending frees as well as inactive capacity.

Side note: the [PyTorch documentation](https://docs.pytorch.org/docs/2.14/generated/torch.Tensor.record_stream.html) warns that allocator polling can race with progress on the recorded stream, making whether a later allocation reuses a block nondeterministic. This raises a separate design question: should MoE replace `record_stream` with explicit event-based stream synchronization? That could avoid dependence on allocator polling while retaining separate streams for communication overlap. It requires auditing buffer lifetimes and synchronization in forward, backward, and CUDA graph capture, then checking correctness and performance independently of the stream-merging experiments below.

**A substantial gap remains between peak active and peak reserved.** These are actual training measurements from rank 0 after #7612, including capture, in GiB:

| Max allocated | Max active | Max reserved | Max reserved minus max active |
| ---: | ---: | ---: | ---: |
| 192.452 | 207.314 | 228.352 | **21.038** |

Most of this remaining gap comes from different allocation streams reaching their active peaks at different times. Cached blocks are associated with their allocation stream, so reservations accumulate even when the corresponding active peaks do not overlap.

The shared graph pool has five allocation streams:

| Graph-pool allocation stream | Peak active (GiB) | Peak reserved (GiB) |
| --- | ---: | ---: |
| Main compute | 169.285 | 169.416 |
| MoE dispatch/combine | 27.939 | 27.965 |
| All-gather | 3.882 | 4.004 |
| Reduce-scatter | 3.208 | 3.223 |
| CP attention | 0.721 | 0.746 |
| **Sum of individual stream peaks** | **205.034** | **205.354** |

```text
sum(peak_active per stream)       = 205.034 GiB
peak_active of the graph pool    = 186.929 GiB
Difference                       =  18.105 GiB
```

The graph pool reserves 205.354 GiB, only 0.320 GiB above the sum of its streams' individual active peaks. Its reserved-minus-active peak gap is therefore 18.425 GiB. An additional 2.613 GiB from the default pool accounts for the full 21.038 GiB gap. These are differences of peaks, not peaks of instantaneous differences.

Running dispatch/combine on the **same main compute stream** could reduce this overhead by allowing their allocations to reuse capacity on that stream. During CUDA graph capture, this means the current capture stream, rather than the literal CUDA default stream. For compute and MoE alone, the sum of their active peaks is 197.223 GiB, while their maximum simultaneous active usage is 183.333 GiB: a **13.891 GiB** difference.

The existing `moe_shortcut_parallel=False` path runs dispatch/combine on the current stream. This is a candidate mitigation, not a measured saving: it also removes their intended overlap with compute. A training comparison should check peak allocated/active/reserved, step time, and loss parity.
